helix-core/src/object.rs

```diff
@@ -59,3 +60,67 @@ pub fn shrink_selection(syntax: &Syntax, text: RopeSlice, selection: &Selection)
         }
     })
 }
+
+pub fn select_next_sibling(syntax: &Syntax, text: RopeSlice, selection: &Selection) -> Selection {
+    let tree = syntax.tree();
+
+    selection.clone().transform(|range| {
+        let from = text.char_to_byte(range.from());
+        let to = text.char_to_byte(range.to());
+
+        let sibling = match tree
+            .root_node()
+            .descendant_for_byte_range(from, to)
+            .and_then(find_next_sibling)
+        {
+            Some(sibling) => sibling,
+            None => return range,
+        };
+
+        let from = text.byte_to_char(sibling.start_byte());
+        let to = text.byte_to_char(sibling.end_byte());
+
+        if range.head < range.anchor {
+            Range::new(to, from)
+        } else {
+            Range::new(from, to)
+        }
+    })
+}
+
+fn find_next_sibling(node: Node) -> Option<Node> {
+    node.next_sibling()
+        .or_else(|| node.parent().and_then(find_next_sibling))
+}
+
+pub fn select_prev_sibling(syntax: &Syntax, text: RopeSlice, selection: &Selection) -> Selection {
+    let tree = syntax.tree();
+
+    selection.clone().transform(|range| {
+        let from = text.char_to_byte(range.from());
+        let to = text.char_to_byte(range.to());
+
+        let sibling = match tree
+            .root_node()
+            .descendant_for_byte_range(from, to)
+            .and_then(find_prev_sibling)
+        {
+            Some(sibling) => sibling,
+            None => return range,
+        };
+
+        let from = text.byte_to_char(sibling.start_byte());
+        let to = text.byte_to_char(sibling.end_byte());
+
+        if range.head < range.anchor {
+            Range::new(to, from)
+        } else {
+            Range::new(from, to)
+        }
+    })
```

Comment:
This whole thing seemed like it is duplicated, probably can reuse.

Reply:
Yep it's all duplicated. I tried out refactoring it to take the `find_prev_sibling`/`find_next_sibling` function as an argument but I couldn't figure out the borrowing rules. I'll keep looking into it :thinking: 

(ofc if you'd like to push a commit to cut down on that boilerplate, I'd be grateful :slightly_smiling_face:)

Reply:
I figured out how to do higher order functions but now I think I may have overdone it :smile: